Add rule files support

loafoe opened the pull request.

loafoe committed:
Add rule files support

loafoe committed:
Update license

loafoe committed:
Backdate the copyright to satisfy check

artemgavrilov commented on go.mod:
```diff
@@ -5,4 +5,5 @@ go 1.14
 require (
 	github.com/stretchr/testify v1.6.1
 	gopkg.in/yaml.v2 v2.3.0
+	gopkg.in/yaml.v3 v3.0.0-20200313102051-9f266ea9e77c
```
Can you please remove `yaml.v2` and leave only `yaml.v3`

loafoe replied:
This change will drop support for `yaml.v2` , but sure

artemgavrilov replied:
yaml.v3 has fallback logic, so old interfaces will work.

https://github.com/go-yaml/yaml/blob/496545a6307b2a7d7a710fd516e5e16e8ab62dbc/decode.go#L421-L424

artemgavrilov commented on rules/groups.go:
```diff
@@ -0,0 +1,43 @@
+	For         string            `yaml:"for,omitempty"`
+	Labels      map[string]string `yaml:"labels,omitempty"`
+	Annotations map[string]string `yaml:"annotations,omitempty"`
+}
```
Let's use `Rule` here instead of `RuleNode` as a reference

https://github.com/prometheus/prometheus/blob/da3fb842b46781801369d63046dc836dae11522f/pkg/rulefmt/rulefmt.go#L114-L122

And for the field `for` let's use type `promconfig.Duration`

artemgavrilov commented on rules/groups_test.go:
```diff
@@ -0,0 +1,55 @@
+		},
+	}
+	data, err := yaml.Marshal(groups)
+	if !assert.Nil(t, err) {
+		return
+	}
```
```suggestion
	require.NoError(t, err)
```

loafoe replied:
Will this stop downstream tests?

palash25 replied:
not sure what your question is but if by "stop downstream tests" you mean will this fail and `return` when it encounters and error then yes

loafoe replied:
Yeah, was not aware of the require construct, will replace

artemgavrilov commented on rules/groups.go:
```diff
@@ -0,0 +1,43 @@
+
+package rules
+
+// Ref: https://github.com/prometheus/prometheus/blob/main/pkg/rulefmt/rulefmt.go
```
Please remove this link, it may become invalid over time. We have links to reference repos in the readme.

loafoe replied:
Done

artemgavrilov commented on rules/groups_test.go:
```diff
@@ -0,0 +1,55 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rules_test
```
Let's use the same package name

```suggestion
package rules
```

artemgavrilov commented on rules/groups.go:
```diff
@@ -0,0 +1,43 @@
+// promconfig
```
Let's rename this file to `rules.go`

loafoe replied:
Done

artemgavrilov commented on rules/groups_test.go:
```diff
@@ -0,0 +1,55 @@
+// promconfig
```
Same, let's rename to `rules_test.go`

loafoe replied:
Done

artemgavrilov reviewed (changes requested):
Thank you for your contribution! We will reuse this config in PMM: https://github.com/percona/pmm-managed/blob/1a54a66408dd3573f0d1da5bb832c1a679cea931/services/management/ia/rules_service.go#L88-L105

loafoe committed:
Changes based on PR feedback:

- Rename test package
- Use `yaml.v3`
- Rename file
- USe `Rule` instead of `RuleNode`

loafoe committed:
Simplify test

artemgavrilov commented on rules/rules.go:
```diff
@@ -0,0 +1,45 @@
+	Rules    []Rule `yaml:"rules"`
+}
+
+// Rule adds yaml.v3 layer to support line and column outputs for invalid rules.
```
```suggestion
// Rule describes an alerting or recording rule.
```

artemgavrilov reviewed (approved):
LGTM except the comment. 👍

It will be nice to have more tests, probably we can copy them from https://github.com/prometheus/prometheus/blob/main/pkg/rulefmt/rulefmt_test.go 
But I can do it myself later.

loafoe committed:
Update rules/rules.go

Co-authored-by: Artem Gavrilov <charlieblackwood7@gmail.com>

artemgavrilov commented on duration.go:
```diff
@@ -117,12 +119,8 @@ func (d Duration) MarshalYAML() (interface{}, error) {
-	var s string
-	if err := unmarshal(&s); err != nil {
-		return err
-	}
-	dur, err := ParseDuration(s)
+func (d *Duration) UnmarshalYAML(node *yaml.Node) error {
+	dur, err := ParseDuration(node.Value)
```
Damn, I was wrong! Sorry, but can you please revert this change and use `yaml.v2` everywhere?

artemgavrilov reviewed (changes requested):
Please rollback to `yaml.v2`, I was wrong regarding it. Also we updated CI in `main` branch, so it makes sense to rebase this branch.

@loafoe ^
